lucet-runtime/lucet-runtime-internals/src/instance/signals.rs

```diff
@@ -174,37 +183,64 @@ extern "C" fn handle_signal(signum: c_int, siginfo_ptr: *mut siginfo_t, ucontext
             }
             SignalBehavior::Terminate => {
                 // set the state before jumping back to the host context
+                //
+                // we must clear the terminable flag so that no KillSwitch accidentally signals a
+                // host thread.
+                inst.kill_state.terminable.store(false, Ordering::SeqCst);
```

Comment:
This `store` and the store added on line 214 are the two most questionable lines of this PR, as far as correctness is concerned, IMO.

I want to prevent the race where _while_ this handler is _finishing_ with a SIGSEGV, some timer decides the instance has timed out, sends a SIGALRM, and ends up sending it to host code. Since this handler masks SIGALRM among others, an alarm raised at or before this point will be blocked. If the instance were terminable (so we got here through a non-alarm fault), exactly one `KillSwitch` may fire, and if it occurs after these `store` that lead to host code, it will fail to get the flag, and not signal.

Because a `KillSwitch` may have fired before (to get us here), or during, this handler, the flag may have been taken, and we can't assert the flag is either `true` or `false`. That leaves just overwriting the flag, and is what makes me question these lines. Because in both cases we will exit to the host, I'm at least confident we won't end up with an unterminable guest.

Reply:
I think it'd help to enumerate the potential interleavings in documentation. It seems like we two variables to explore: currently handling signal raised by a `KillSwitch` or no? `KillSwitch` signal arrives before or after this `store`?